src/Microsoft.OData.Core/Uri/NodeToStringBuilder.cs

```diff
@@ -470,7 +471,240 @@ internal String TranslateNode(QueryNode node)
+        {
+            switch(node.Kind)
+            {
+                case QueryNodeKind.SingleValuePropertyAccess:
+                    SingleValuePropertyAccessNode singleValuePropertyAccessNode = node as SingleValuePropertyAccessNode;
+                    return GetResourceRangeVariableReferenceNode(singleValuePropertyAccessNode.Source);
+
+                case QueryNodeKind.Convert:
+                    ConvertNode convertNode = node as ConvertNode;
+                    return GetResourceRangeVariableReferenceNode(convertNode.Source);
+
+                case QueryNodeKind.Any:
+                    AnyNode anyNode = node as AnyNode;
+                    return GetResourceRangeVariableReferenceNode(anyNode.Source);
+
+                case QueryNodeKind.SingleValueFunctionCall:
+                    SingleValueFunctionCallNode singleValueFunctionCallNode = node as SingleValueFunctionCallNode;
+                    return GetResourceRangeVariableReferenceNode(singleValueFunctionCallNode.Parameters.First());
+
+                case QueryNodeKind.ResourceRangeVariableReference:
+                    return node as ResourceRangeVariableReferenceNode;
+
+                case QueryNodeKind.SingleValueOpenPropertyAccess:
+                    SingleValueOpenPropertyAccessNode singleValueOpenPropertyAccessNode = node as SingleValueOpenPropertyAccessNode;
+                    return GetResourceRangeVariableReferenceNode(singleValueOpenPropertyAccessNode.Source);
+
+                case QueryNodeKind.SingleComplexNode:
+                    SingleComplexNode singleComplexNode = node as SingleComplexNode;
+                    return GetResourceRangeVariableReferenceNode(singleComplexNode.Source);
+
+                case QueryNodeKind.CollectionComplexNode:
+                    CollectionComplexNode collectionComplexNode = node as CollectionComplexNode;
+                    return GetResourceRangeVariableReferenceNode(collectionComplexNode.Source);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Bind $it to the <see cref="QueryNode"/> translated string.
+        /// </summary>
+        /// <param name="node">node to bind.</param>
+        /// <param name="filterClauseRangeVariable">The <see cref="FilterClause"/> range variable.</param>
+        /// <returns>The translated string with $it binding.</returns>
+        private string BindNode(QueryNode node, ResourceRangeVariable filterClauseRangeVariable)
+        {
+            BinaryOperatorNode binaryNode = node as BinaryOperatorNode;
+            InNode inNode = node as InNode;
+            AnyNode anyNode = node as AnyNode;
+            SingleValueFunctionCallNode singleValueFunctionCallNode = node as SingleValueFunctionCallNode;
+            SingleValueOpenPropertyAccessNode singleValueOpenPropertyAccessNode = node as SingleValueOpenPropertyAccessNode;
+
+            if (binaryNode != null)

```

Comment:
Alternatively,

```csharp
BinaryOperatorNode binaryNode;
// ... other variable declarations

if ((binaryNode = node as BinaryOperatorNode) != null)
{
   // ...
}
// else if ...
```
I however don't know if the cost of doing all 5 explicit conversions using the `as` operator regardless is significant. In this case it might not matter since the most common scenario is handled by the `else` block and all the explicit conversions will have been done either way.

Reply:
Let me hear the opinions of others on this.

Reply:
I had the same thought -- we want to limit the number of casts.

Can we switch on QueryNode.Kind (or QueryNode.InternalKind)?

---
In reply to: [574336006](https://github.com/OData/odata.net/pull/1991#discussion_r574336006) [](ancestors = 574336006)